### ISSUE TITLE
Supporting loops in scripts

### ISSUE THREAD

**Issue by angelhof** [binpash/hs#36]
## Feature Proposal: Loops

This issue describes what we need to support loops in shell programs. Currently, our scheduler only handles known partial orders where everything is unrolled and each node is supposed to execute once. In real programs however, we have loops (and even nested loops) where we can't (always) determine the number of times they will execute. This issue describes a conservative solution to resolve this issue and some hints on how to make this solution more optimistic to allow for more parallelism.

### Example loop

```sh
for i ...
    cmd1
    for j ...
        cmd2
        cmd3
    done
    cmd4
done
```

### Proposal: Loop nodes in partial order

To support these nodes we could extend partial orders to abstract partial orders that also contain loop nodes, i.e., nodes that have an associated stack of loop identifiers. 
The abstract partial order for the above program is shown in the attached picture.

![image0(6)](https://user-images.githubusercontent.com/12186758/234633250-9ee308ab-a96a-4b02-a639-a2d79f6821d3.jpeg)

Each loop can then be extended with a loop identifier in the preprocessing pass (kk: is this sound) and everytime the JIT sends a wait to the scheduler it also passes the stack of loop contexts (node_id.loop1.loop2) to uniquely idenitfy the abstract loop node as well as a loop iteration counter for each loop (to uniquely identify the instance of the abstract loop node).

Instances of abstract loop nodes are ordered syntactically by doing an alpha-sort of their stack of iteration counts.

### Scheduling

To start with and incrementally support loops, I propose starting execution of a loop node instance only after we receive the wait for now (non-speculative execution for loop nodes). Later down the line, we will be able to speculate loop nodes by guessing the next variable values, but we should not start with that.

The scheduler then knows that a loop is done iterating when it receives the wait for a node after the loop. Note that this doesn't mean that the loop is completely done, e.g., when having nested loops. Later we could consider sending a loop-done message to the scheduler right after the loop is done.

**Comment by mgree** [comment]
One big advantage here is that `for` loops operate by fully expanding their iterees before running:

```sh
for x in words...
do
  cmd
  ...
done
```

will _first_ fully expand `words` and then iterate through it based on the `IFS` value at the start of the for loop. The body of the loop can change things with `break`, `continue`, `return`, and `exit`, but it can't modify what's being iterated. So we can't determine exactly how many times they execute, but we _can_ put an upper bound on it. So we can start with the pessimistic, "all iterations run" schedule and then just watch for control changes.

**Comment by angelhof** [comment]
Very good point! So I guess that it should be possible to have some speculative execution of different loop iterations at the same time. 

I guess that this does not hold for while and until loops so we will probably have to address them separately if needed :)

**Comment by mgree** [comment]
Right---while/until loops just test the condition each time. We can use
information on a condition's reads/writes the first time to try to
schedule it the next run, but we shouldn't _ever_ speculate on a loop
condition until at least one of its reads (incl. environment variables)
has changed.

**Comment by angelhof** [comment]
Conservative loop unrolling addressed (for now) by PR #45 . Let's open a new issue when we want to think about more optimistic unrolling
